### DISCUSSION TITLE
Per-crate fileset

### DISCUSSION THREAD

**Discussion by mightyiam** [90-008/nix-cargo-integration#166] [Q&A] [answered]
Different crates within a project can depend on different non-Rust files (compile-time assets).

Practically, we probably want a single source derivation for the entire project. And I'm guessing that's what nci does, right?

What I would enjoy is a way to specify inclusions (fileset library set?) per crate, and for nci to somehow compose them for us using option merging.

This is for separation of concerns.

Does this make sense?

**Comment by 90-008** [comment]
There is a new fileset API in nixpkgs already https://nixos.org/manual/nixpkgs/unstable/#sec-fileset so i don't really see myself implementing anything in NCI, one could simply just use this, and pass it to `projects.<name>.path`.

(it might be useful to include an example for it though?)

**Reply by mightyiam** [reply]
Sorry, I must have not explained myself sufficiently. I tried what you've described and obviously it works, but I don't find it entirely satisfying, because I would like more separation of concerns.

But maybe I'm acting on a wrong assumption. If I don't specify individual crates, but a project, do the crates end up sharing the same source package?

**Reply by 90-008** [reply]
the only source that NCI will use is the source you provide it in `projects.<name>.path`. if you meant by separation of concerns having a per-crate option, this doesn't make sense here IMO.

**Reply by mightyiam** [reply]
So, one crate is `include_bytes!` some checked in file, and another crate is `include_str!` some other. And I have a flake-parts module for each of them. Where should the filtering code go?

**Reply by 90-008** [reply]
you should be able to just override the crate derivation by setting `mkDerivation.src = lib.mkForce my-src`, i'm not completely sure if this works but it should in theory. that's what i'd rather "support" if it doesn't already work

**Reply by mightyiam** [reply]
Hrm... but then I end up with more duplication in the store. I can always invent my own options...

**Reply by 90-008** [reply]
that is true, but the duplication is unavoidable whether or not an option exists, because the project path always has to be passed, since it is used for discovering crates and whatnot

**Reply by mightyiam** [reply]
Sorry, what do you mean by "passed", please? Passed where?

**Reply by 90-008** [reply]
the project path option is what i am talking about. it has to be used so that nci can discover the crates. if you wanted to have different sources for different crates, it would be duplicated, and you still need to keep in mind crates will most likely depend on each other in a monorepo (afterall that is the point of it) so is it really worth separating the sources?

**Reply by mightyiam** [reply]
I don't want different sources for different crates. I think I want each create to define its own fileset and they will all be merged together with `lib.fileset.unions`. This way, there's both separation of concerns and a single source.

**Reply by 90-008** [reply, marked as answer]
there could definitely be an option for this but i don't really know if it's worth adding in nci when it would be a really simple option in the first place, it doesn't even need to touch any nci code. personally i'd rather if it was a seperate flake-parts module that's not in nci, i think it's outside the concerns for nci

**Reply by mightyiam** [reply]
[Here is what we ended up doing](https://github.com/molybdenumsoftware/pr-tracker/commit/d6e006e7370f143c769b0085f4d50704a274d69e#diff-ed4806b080fd67fe8134af3e47fa35b64f4bb588178374f48de51217e4bcf68bR14).

Each module can specify its own fileset and they are merged using unions.

There's a bit more detail that you can observe.
